businessRules/donations.js

```diff
@@ -1,54 +1,62 @@
 const { v4 } = require('uuid');
 const { createDonation } = require('../models/donations');
-const { getAllIdeas } = require('../models/ideas');
-const { getIdeas } = require('../utils/storage');
+const { findIdea } = require('../validations/ideas');
+const {
+    validateSum,
+    findIdeaWithTotalDonationSum,
+} = require('../validations/donations.js');
+const { findUser } = require('../validations/users');
 
-function createDonation(firstName, sum, userId, ideaId) {
-    let ideas = getIdeas();
-    const idea = ideas.find((idea) => ideaId === idea.id);
+function createDonationByUnregisteredUser(sum, userId, ideaId) {
+    findIdea(ideaId);
+    validateSum(sum);
 
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
+    const user = findUser(userId);
+
+    let firstName;
+
+    if (!user) {
+        const generateId = v4();
+        const anonymous = 'Anonymous';
+
+        firstName = firstName || anonymous;
+        userId = userId || generateId;
     }
+    createDonation(firstName, sum, userId, ideaId);
+}
 
-    const generateId = v4();
-    const anonymous = 'Anonymous';
+function createDonationByRegisteredUser(sum, userId, ideaId) {
+    validateSum(sum);
 
-    firstName = firstName || anonymous;
-    userId = userId || generateId;
+    const user = findUser(userId);
+
+    const firstName = user.firstName;
 
     createDonation(firstName, sum, userId, ideaId);
 }
 
-function ideasDonationSum(ideaId) {
-    const ideas = getAllIdeas();
-    const idea = ideas.find((idea) => idea.id === ideaId);
+function getIdeasTotalDonationSum(ideaId) {
+    const idea = findIdeaWithTotalDonationSum(ideaId);
```

Comment:
I would rename `findIdeaWithTotalDonationSum` to `getIdeaWithTotalDonationSum`, because `find` suggests that you're searching and you might not find it, while the expectation here is that you will get a valid idea id, so you're actually simply getting an idea by idea id

Reply:
Oh, OK. If it returns something- it GETs! Well in this certain situation. )